lib/segment/src/index/hnsw_index/hnsw.rs

```diff
@@ -219,75 +220,80 @@ impl<TGraphLinks: GraphLinks> HNSWIndex<TGraphLinks> {
+        // - `params` is `Some`
+        // - `params.quantization` is `Some`
+        // - and `params.quantization.ignore` is `false`
+        let quantization_params =
+            params
+                .and_then(|p| p.quantization)
+                .unwrap_or_else(|| QuantizationSearchParams {
+                    ignore: default_quantization_ignore_value(),
+                    rescore: default_quantization_rescore_value(),
+                    oversampling: default_quantization_oversampling_value(),
+                });
+
+        let (raw_scorer, quantized) = match quantized_storage {
+            // If `quantization_params` is `Some`, then quantization is *not* ignored
+            Some(quantized_storage) if !quantization_params.ignore => {
+                let scorer = quantized_storage.raw_scorer(
                     vector,
                     id_tracker.deleted_point_bitslice(),
                     vector_storage.deleted_vector_bitslice(),
-                ),
-                true,
-            )
-        } else {
-            (
-                new_raw_scorer(
+                );
+
+                (scorer, true)
+            }
+
+            _ => {
+                let scorer = new_raw_scorer(
                     vector.to_owned(),
                     &vector_storage,
                     id_tracker.deleted_point_bitslice(),
-                ),
-                false,
-            )
+                );
+
+                (scorer, false)
+            }
         };
-        let payload_index = self.payload_index.borrow();
 
+        let payload_index = self.payload_index.borrow();
         let filter_context = filter.map(|f| payload_index.filter_context(f));
-
         let points_scorer = FilteredScorer::new(raw_scorer.as_ref(), filter_context.as_deref());
 
-        if let Some(graph) = &self.graph {
-            let search_result = graph.search(top, ef, points_scorer);
-            let if_rescore = params
-                .and_then(|p| p.quantization)
-                .map(|q| q.rescore)
-                .unwrap_or(default_quantization_rescore_value());
-            if quantized && if_rescore {
-                let raw_scorer = new_raw_scorer(
-                    vector.to_owned(),
-                    &vector_storage,
-                    id_tracker.deleted_point_bitslice(),
-                );
-                let mut ids_iterator = search_result.iter().map(|x| x.idx);
-                let mut re_scored = raw_scorer.score_points_unfiltered(&mut ids_iterator);
-                re_scored.sort_unstable();
-                re_scored
-            } else {
-                search_result
-            }
+        let Some(graph) = &self.graph else {
+            return Vec::new();
+        };
+
+        if quantized && quantization_params.rescore {
+            let oversampled_top = quantization_params
+                .oversampling
+                .map(|x| (top as f64 * max(x, 1.0)) as usize)
+                .unwrap_or(top);
```

Comment:
What about this? More verbose, but easier to read, IMO.  🤔

```suggestion
            let oversampling = quantization_params
                .oversampling
                .unwrap_or(1.0);

            // Apply oversampling if necessarry (and ignore oversampling values less than `1.0`)
            let oversampled_top = if oversampling > 1.0 {
                (oversampling * top as f64) as usize
            } else {
                top
            };
```

Reply:
> What about this?

I prefer `if` blocks. `if`s are trivial to read, and `bool::then_some` is a form of an IQ test, IMO. 🙈